### ISSUE TITLE
EDD()->payment_stats->get_earnings Returns inaccurate totals when new orders are created.

### ISSUE THREAD

**Issue by alexstandiford** [easydigitaldownloads/easy-digital-downloads#7601]
### Expected behavior

I expect that `EDD()->payment_stats->get_earnings()` would return an accurate total of earnings when a purchase was recently made.

### Actual behavior
If a purchase was recently made, `EDD()->payment_stats->get_earnings()` does not account for this purchase.

### Steps to reproduce the behavior
1. View the result of `EDD()->payment_stats->get_earnings()` (this ensures the transient is created, more on that later)
1. Place an order using EDD
1. View the result of `EDD()->payment_stats->get_earnings()` again
1. Open the earnings reports, and set the filter to "this month"
1. Observe that the reported earnings do not match the earnings depicted in `EDD()->payment_stats->get_earnings()`.

### Information (if a specific version is affected):
PHP Version: 7.3.2

EDD Version (or branch): 2.9.20

WordPress Version: 5.3.2

#### Any other relevant information:

This appears to be caused by a transient that does not get wiped when a payment is created. In my testing environment, I created a few orders, and here's my findings:

here's what EDD calculates for all orders for this month:
![image](https://user-images.githubusercontent.com/8210827/75374196-cd8c8380-5899-11ea-88bc-806bb2b046ef.png)

However, when I run `EDD()->payment_stats->get_earnings()`, I get `2783.29`.

When I run `wp transient get edd_stats_earnings` in WP CLI, I get an array of serialized queries and their corresponding values.

```
wp transient get edd_stats_earnings

array (
  '8a41c76997470b4b576a96eec523a377' => 2783.2899999999991,
  'e082ecc77d299d624ba98fe562d10d99' => 256.07999999999998,
  '7bdbfa543cd7b8a3cf94c3f14f6f62a9' => 2360.0999999999995,
)
```

If I clear this transient:

```
wp transient delete edd_stats_earnings
Success: Transient deleted.
```

And run `EDD()->payment_stats->get_earnings()` again, I get the expected `2979.19`.

**Comment by cklosowski** [comment]
This is our cached response, as it's a heavy call. As far as I can tell
this is expected for now. 3.0 solves a lot of this due to the fact that we
can run SUM queries and the cache invalidation is handled for us in Berlin.

Unless I'm missing something.

**Comment by alexstandiford** [comment]
Yep, that totally makes sense, however, this cache causes the results to be perpetually 1-hour behind the actual amounts, as that's when the transient clears. To get around this, we have had to bypass the transient altogether:

```php
<?php

// Short circuit the EDD cache
add_filter( 'pre_transient_edd_stats_earnings', '__return_empty_array' );

$date_range = $this->prepare_date_range( $date );

$result = (float) EDD()->payment_stats->get_earnings();

remove_filter( 'pre_transient_edd_stats_earnings', '__return_empty_array' );
?>
```

Is there any reason why this transient is not reset when a new purchase is made? I think that would keep the records accurate while also benefitting from the cache.

**Comment by alexstandiford** [comment]
I suppose the only queries that are impacted are the ones that _include current data_, so if there was a way to only remove the cached totals from queries that include today's date, you could probably get the best of both worlds, and keep all older queries without causing current queries to be inaccurate.

**Comment by cklosowski** [comment]
The main reason this transient isn’t cleared when a new purchase is made is that this transient would be rewritten a large number of times per day, essentially making this transient ignored most of the time as it would be wiped out.

After discussing with Alex we’re going to close this as it is solved within EDD 3.0 a bit more elegantly. This also leaves us with a record for the history and if we want to re-open it later for any other solutions we can.
